crates/holochain_persistence_lmdb/src/eav/lmdb.rs

```diff
@@ -0,0 +1,299 @@
+use holochain_persistence_api::{
+    cas::content::AddressableContent,
+    eav::{
+        Attribute, EavFilter, EaviQuery, EntityAttributeValueIndex, EntityAttributeValueStorage,
+    },
+    error::{PersistenceError, PersistenceResult},
+    reporting::{ReportStorage, StorageReport},
+};
+// use kv::{Config, Manager, Store, Error as KvError};
+use rkv::{
+    error::StoreError, DatabaseFlags, EnvironmentFlags, Manager, Rkv, SingleStore, StoreOptions,
+    Value,
+};
+use std::{
+    collections::BTreeSet,
+    fmt::{Debug, Error, Formatter},
+    marker::{PhantomData, Send, Sync},
+    path::Path,
+    sync::{Arc, RwLock},
+};
+use uuid::Uuid;
+
+const EAV_BUCKET: &str = "EAV";
+const MAX_SIZE_BYTES: usize = 104857600; // TODO: Discuss what this should be, currently 100MB
+
+#[derive(Clone)]
+pub struct EavLmdbStorage<A: Attribute> {
+    id: Uuid,
+    store: SingleStore,
+    manager: Arc<RwLock<Rkv>>,
+    attribute: PhantomData<A>,
+}
+
+impl<A: Attribute> EavLmdbStorage<A> {
+    pub fn new<P: AsRef<Path> + Clone>(db_path: P) -> EavLmdbStorage<A> {
+        let cas_db_path = db_path.as_ref().join("eav").with_extension("db");
+        std::fs::create_dir_all(cas_db_path.clone())
+            .expect("Could not create file path for CAS store");
+
+        let manager = Manager::singleton()
+            .write()
+            .unwrap()
+            .get_or_create(cas_db_path.as_path(), |path: &Path| {
+                let mut env_builder = Rkv::environment_builder();
+                env_builder
+                    // max size of memory map, can be changed later
+                    .set_map_size(MAX_SIZE_BYTES)
+                    // max number of DBs in this environment
+                    .set_max_dbs(1)
+                    // Thes flags make writes waaaaay faster by async writing to disk rather than blocking
+                    // There is some loss of data integrity guarantees that comes with this
+                    .set_flags(EnvironmentFlags::WRITE_MAP | EnvironmentFlags::MAP_ASYNC);
+                Rkv::from_env(path, env_builder)
+            })
+            .expect("Could not create the environment");
+
+        let env = manager
+            .read()
+            .expect("Could not get a read lock on the manager");
+
+        // Then you can use the environment handle to get a handle to a datastore:
+        let options = StoreOptions {
+            create: true,
+            flags: DatabaseFlags::empty(),
+        };
+        let store: SingleStore = env
+            .open_single(EAV_BUCKET, options)
+            .expect("Could not create EAV store");
+
+        EavLmdbStorage {
+            id: Uuid::new_v4(),
+            store: store,
+            manager: manager.clone(),
+            attribute: PhantomData,
```

Comment:
I wonder if we can just DRY out this code somewhere just so that it doesnt live in two places?